README.md
# jmbus
This is a non-authorative fork of the Java M-Bus library jMBus / https://www.openmuc.org/m-bus/ by the openMUC project of the Department “Intelligent Energy Systems” at the Fraunhofer Institute for Solar Energy Systems in Freiburg, Germany. Please refer to their website!

jMBus is licensed under the GPLv3 and so are the changes.

This fork has only been started to introduce some additions to decode proprietary messages sent by Techem heat cost allocator devices (Heizkostenverteiler) via Wireless M-Bus. Those changes are submitted upstream, it is unsure, wheter they will be included or not.


doc/CHANGELOG.txt
v2.2.2 04-Jul-2016
------------------
- added support for negative BCDs thanks to Mr Pirscoveanu from
  neogrid
- fixed bug receiving a second wireless M-Bus message right aftern
  another.
- added support for additional units from the newest IEC 62056-6-2
  standard
- aded gradle wrapper scripts to distribution so that gradle
  installation is not needed

v2.2.1 26-Feb-2016
------------------
- increased default timeout from 300ms to 500ms because it was too
  short for many meters
- made timeout configurable in all console applications
- inserted sleep between SND_NKE and REQ_UD2 requests in ReadMeter and
  ScanMeter console apps because some meters would otherwise discard
  the REQ_UD2 request
- fixed decoding subunit, tariff and storage number from DIFEs
- added support for decoding new data records defined in EN
  13757-3-2013

v2.2.0 18-Nov-2015
------------------
- added support for writing to meters
- added a command-line application for writing to meters
- added support for scanning secondary addresses
- refactored decoding function to improve performance
- added support for Float32 decoding
- several minor improvements
- added possibility to select certain records/registers for readout
- changed from java 6 to java 7

v2.1.0 21-Jan-2015
------------------
- added support for secondary addressing using wired M-Bus

v2.0.0 02-Dec-2014
------------------
- changed license from LGPL to GPLv3 or higher
- added support for wireless M-Bus. Transceivers supported are from Amber
  and RadioCrafts
- API clean up, for better understanding
- added detailed user guide on how M-Bus works and how jMBus is used.

v1.3 17-Jul-2014
----------------
- added decoding support for more VIBs and DIBS
- fixed build script to work with newest gradle
- fixed support for two digit BCDs
- BCD now extends class Number again
- added run script for ScanForMeters app
- added thorough description on how to get RXTX running
- RXTX src + native libs are now included in the distribution
- added documentation on how to compile your own nativ part of RXTX.
- added integration tests for better testing

v1.2 16-Jan-2014
----------------
- ScanForMeter application added that can be used to scan all primary
  addresses for meters
- fix parsing manufacturer id
- added decoding of alternate exteded VIFE field
- ReadMeter application now prints more valuable information

v1.1.1 09-Jan-2014
------------------
- reinserted TimeoutException when reading
- releasing to maven central repository

v1.0.0 07-Jan-2014
------------------
- bug fix parsing manufacturer specific data
- simple ReadMeter application is now part of the library. Its usage
  is documented in README.txt
- serialports can be closed and opened again through the API
- DLMSUnit is now an Enum

v0.10.0 04-Sep-2013
-------------------
- read now supports timeouts
- serial port settings string is correctly parsed
- data type can now be retrieved from DIF/VIF
- finding correct serial port is now more efficient
- bug fix in decoding DIF/VIF

v0.9.8 15-Mar-2012
------------------
- changed build system to gradle
- modified documentation and usage example

v0.9.7 20-Oct-2011
------------------
- short message is send in one command in MBusLSAP to prevent timeout,
  thanks to Hakan Gonene

v0.9.6 19-Sep-2011
------------------
- added additional data types: data, date/time, varlen string
- moved to maven

v0.9.4 3-Dec-2010
-----------------
- Fixed build.xml file so a modfied library can be easily build.
- Fixed some bugs when parsing M-Bus messages

v0.9.3 17-Nov-2010 
------------------
- Fixed bug with integer treatment
- Implemented application reset command
- added voltage and current VIB interpretation
- Debug output switch with org.openmuc.jmbus.debug system property


src/main/java/org/openmuc/jmbus/DlmsUnit.java
/*
 * Copyright 2010-16 Fraunhofer ISE
 *
 * This file is part of jMBus.
 * For more information visit http://www.openmuc.org
 *
 * jMBus is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * jMBus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with jMBus.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.openmuc.jmbus;

import java.util.HashMap;
import java.util.Map;

/**
 * The units as defined in IEC 62056-6-2. Some units not defined in IEC 62056-6-2 but needed by M-Bus were added.
 * 
 */
public enum DlmsUnit {
    // can be found in IEC 62056-6-2 2013 Capture 5.2.2
    YEAR(1, "a"),
    MONTH(2, "mo"),
    WEEK(3, "wk"),
    DAY(4, "d"),
    HOUR(5, "h"),
    MIN(6, "min"),
    SECOND(7, "s"),
    DEGREE(8, "°"),
    DEGREE_CELSIUS(9, "°C"),
    CURRENCY(10, ""),
    METRE(11, "m"),
    METRE_PER_SECOND(12, "m/s"),
    CUBIC_METRE(13, "m³"),
    CUBIC_METRE_CORRECTED(14, "m³"),
    CUBIC_METRE_PER_HOUR(15, "m³/h"),
    CUBIC_METRE_PER_HOUR_CORRECTED(16, "m³/h"),
    CUBIC_METRE_PER_DAY(17, "m³/d"),
    CUBIC_METRE_PER_DAY_CORRECTED(18, "m³/d"),
    LITRE(19, "l"),
    KILOGRAM(20, "kg"),
    NEWTON(21, "N"),
    NEWTONMETER(22, "n"),
    PASCAL(23, "Nm"),
    BAR(24, "bar"),
    JOULE(25, "J"),
    JOULE_PER_HOUR(26, "J/h"),
    WATT(27, "W"),
    VOLT_AMPERE(28, "VA"),
    VAR(29, "var"),
    WATT_HOUR(30, "Wh"),
    VOLT_AMPERE_HOUR(31, "VAh"),
    VAR_HOUR(32, "varh"),
    AMPERE(33, "A"),
    COULOMB(34, "C"),
    VOLT(35, "V"),
    VOLT_PER_METRE(36, "V/m"),
    FARAD(37, "F"),
    OHM(38, "Ohm"),
    OHM_METRE(39, "Ohm m²/m"),
    WEBER(40, "Wb"),
    TESLA(41, "T"),
    AMPERE_PER_METRE(42, "A/m"),
    HENRY(43, "H"),
    HERTZ(44, "Hz"),
    ACTIVE_ENERGY_METER_CONSTANT_OR_PULSE_VALUE(45, "1/(Wh)"),
    REACTIVE_ENERGY_METER_CONSTANT_OR_PULSE_VALUE(46, "1/(varh)"),
    APPARENT_ENERGY_METER_CONSTANT_OR_PULSE_VALUE(47, "1(VAh)"),
    VOLT_SQUARED_HOURS(48, "V²h"),
    AMPERE_SQUARED_HOURS(49, "A²h"),
    KILOGRAM_PER_SECOND(50, "kg/s"),
    KELVIN(52, "S"),
    VOLT_SQUARED_HOUR_METER_CONSTANT_OR_PULSE_VALUE(53, "K"),
    AMPERE_SQUARED_HOUR_METER_CONSTANT_OR_PULSE_VALUE(54, "1/(V²h)"),
    METER_CONSTANT_OR_PULSE_VALUE(55, "1/(A²h)"),
    PERCENTAGE(56, "%"),
    AMPERE_HOUR(57, "Ah"),

    ENERGY_PER_VOLUME(60, "Wh/m³"),
    CALORIFIC_VALUE(61, "J/m³"),
    MOLE_PERCENT(62, "Mol %"),
    MASS_DENSITY(63, "g/m³"),
    PASCAL_SECOND(64, "Pa s"),
    SPECIFIC_ENERGY(65, "J/kg"),

    SIGNAL_STRENGTH(70, "dBm"),

    RESERVED(253, ""),
    OTHER_UNIT(254, ""),
    COUNT(255, ""),
    // not mentioned in 62056, added for MBus:
    CUBIC_METRE_PER_SECOND(150, "m³/s"),
    CUBIC_METRE_PER_MINUTE(151, "m³/min"),
    KILOGRAM_PER_HOUR(152, "kg/h"),
    CUBIC_FEET(153, "cft"),
    US_GALLON(154, "Impl. gal."),
    US_GALLON_PER_MINUTE(155, "Impl. gal./min"),
    US_GALLON_PER_HOUR(156, "Impl. gal./h"),
    DEGREE_FAHRENHEIT(157, "°F");

    private final int id;
    private final String unit;

    private static final Map<Integer, DlmsUnit> idMap = new HashMap<Integer, DlmsUnit>();

    static {
        for (DlmsUnit enumInstance : DlmsUnit.values()) {
            if (idMap.put(enumInstance.getId(), enumInstance) != null) {
                throw new IllegalArgumentException("duplicate ID: " + enumInstance.getId());
            }
        }
    }

    private DlmsUnit(int id, String unit) {
        this.id = id;
        this.unit = unit;
    }

    /**
     * Returns the ID of this DlmsUnit.
     * 
     * @return the ID
     */
    public int getId() {
        return id;
    }

    /**
     * Returns the unit sign of this DlmsUnit.
     * 
     * @return the ID
     */
    public String getUnit() {
        return unit;
    }

    /**
     * Returns the DlmsUnit that corresponds to the given ID. Returns DlmsUnit.RESERVED if no DlmsUnit with the given ID
     * exists.
     * 
     * @param id
     *            the ID
     * @return the DlmsUnit that corresponds to the given ID
     */
    public static DlmsUnit getInstance(int id) {
        DlmsUnit enumInstance = idMap.get(id);
        if (enumInstance == null) {
            enumInstance = DlmsUnit.RESERVED;
        }
        return enumInstance;
    }

}
